build-tools/packages/build-cli/src/library/packageExports.ts

```diff
@@ -0,0 +1,214 @@
+			return { value };
+		}
+	}
+	return undefined;
+}
+
+// Some common "exports" conditions
+const typesExportCondition = "types";
+const defaultExportCondition = "default";
+
+/**
+ * Performs a depth first search of exports conditions looking for a "types" constrained
+ * resolution path (relative file path) matching keys in query map.
+ *
+ * @param mapQueryPathToOutKey - map with match keys
+ * @param exports - export conditions
+ * @param previous - accumulated conditions and sense of t
+ * @returns matching export data and matching query entry value as outKey property
+ */
+function findFirstTypesPathMatching<TOutKey>(
+	mapQueryPathToOutKey: ReadonlyMap<string | RegExp, TOutKey | undefined>,
+	exports: Readonly<ExportsRecordValue>,
+	previous: Readonly<{
+		conditions: readonly string[];
+		isTypeOnly?: boolean;
+	}>,
+): (ExportData & { outKey: TOutKey | undefined }) | undefined {
+	// All exports are type only if there was a previous condition where the only option
+	// was "types" constrained. Otherwise they may still become constrained or not.
+	const isTypeOnlySettled =
+		(previous.isTypeOnly ?? false) && previous.conditions.includes(typesExportCondition);
+	const entries = Object.entries(exports);
+	for (const [entry, value] of entries) {
+		// Current conditions
+		// "default" is not an explicit condition, but a catch all; so, never add it to conditions
+		const conditions =
+			entry === defaultExportCondition ? previous.conditions : [...previous.conditions, entry];
+		const isTypeOnly =
+			isTypeOnlySettled || (entries.length === 1 && entry === typesExportCondition);
+		// First check if this entry is a leaf; where value is only
+		// expected to be a string (a relative file path).
+		if (typeof value === "string") {
+			const relPath = value;
+			// At the leaf level, look for "types" entries which either is the current
+			// condition (entry) or is an inherited condition, both of which have been
+			// combined into local conditions.
+			if (conditions.includes(typesExportCondition)) {
+				const queryResult = valueOfFirstKeyMatching(relPath, mapQueryPathToOutKey);
+				if (queryResult !== undefined) {
+					return { outKey: queryResult.value, relPath, conditions, isTypeOnly };
+				}
+			}
+		} else if (value !== null) {
+			// Type constrain away array set that is not supported (and not expected
+			// but non-fatal to known use cases).
+			if (Array.isArray(value)) {
+				continue;
+			}
+			const deepFind = findFirstTypesPathMatching(mapQueryPathToOutKey, value, {
+				conditions,
+				isTypeOnly,
+			});
+			if (deepFind !== undefined) {
+				return deepFind;
+			}
+		}
+	}
+
+	return undefined;
+}
+
+/**
+ * Read package "exports" to determine which of given file paths are present.
+ *
+ * @param packageJson - json content of package.json
+ * @param mapQueryPathToOutKey - keys of map represent paths to match. When matched
+ * value, if defined, is used to set entry key in output mapKeyToOutput.
```

Comment:
Took me a couple of reads to convince myself I had (hopefully) understood this one right.
```suggestion
 * @param mapQueryPathToOutKey - keys of map represent paths to match. When one of those paths is
 * found in the package.json's exports, the corresponding value (if defined) is used to set entry key in output mapKeyToOutput.
```